src/app.md

```diff
@@ -251,83 +251,38 @@ fn run() -> Result<()> {
 
 [ex-find-file-loops]: #ex-find-file-loops
 <a name="ex-find-file-loops"></a>
-## Find loops within a directory
+## Find loops for a given path
 
 [![same_file-badge]][same_file] [![cat-filesystem-badge]][cat-filesystem]
 
-Use [`same-file::is_same_file`] to detect loops within a directory.
-This manually iterates through the filesystem via the [`std::path`] module.
+Use [`same-file::is_same_file`] to detect loops for a given path.
 For example, a loop could be created on a unix system via sym links:
 ```
 mkdir -p /tmp/foo/bar/baz
 ln -s /tmp/foo/  /tmp/foo/bar/baz/qux
 ```
-Then, running the following would detect it and print that there's a loop between
-`/tmp/foo` and `/tmp/foo/bar/baz/qux`
+Then, running the following would assert that there exists a loop.
 
 ```rust,no_run
 extern crate same_file;
 
 use std::io;
-use std::fs;
-use std::path::{Path, PathBuf};
-use std::collections::HashMap;
+use std::path::Path;
 use same_file::is_same_file;
 
-// Recursively visit directories until we run out or find a loop
-fn find_loop(
-    dir: &Path,
-    root: &Path,
-    file_tree: &mut HashMap<PathBuf, PathBuf>,
-) -> io::Result<Option<(PathBuf, PathBuf)>> {
-    if dir.is_dir() {
-        for entry in fs::read_dir(dir)? {
-            let path = entry?.path();
-            file_tree.insert(path.clone(), dir.to_path_buf());
-            let loop_check = check_for_loop(&path, &root, file_tree)?;
-            if let Some(loop_files) = loop_check {
-                return Ok(Some(loop_files));
-            } else if path.is_dir() {
-                if let Some(loop_files) = find_loop(&path, root, file_tree)? {
-                    return Ok(Some(loop_files))
-                }
-            }
-        }
-    }
-    Ok(None)
-}
-
-// Given a path, compare the path with all of its parents to find a loop
-// The Hashmap will look something like:
-// "/usr/bin" -> "/usr", "/usr/share" -> "/usr", "/usr" -> "/", "/" -> "/"
-fn check_for_loop(
-    path: &Path,
-    root: &Path,
-    file_tree: &HashMap<PathBuf, PathBuf>,
-) -> io::Result<Option<(PathBuf, PathBuf)>> {
-    let mut parent = &file_tree[path];
-    loop {
-        // Check to see if path is the same as one of its parents
-        if is_same_file(path, parent)? {
-            return Ok(Some((path.to_owned(), parent.to_owned())));
-        } else if is_same_file(parent, root)? {
-            // We've checked all the way to the root
-            return Ok(None);
-        } else {
-            // Check the next parent
-            parent = &file_tree[parent];
+// Check this path against all of its parents
+fn contains_loop(path: &Path) -> io::Result<bool> {
+    let mut path_buf = path.to_path_buf();
+    while path_buf.pop() {
+        if is_same_file(&path_buf, path)? {
+            return Ok(true);
```

Comment:
- the contains_loop tests only if the whole path is pointing to one of its ancestors while we would like to find all loops. calling it repeatedly on poped path segments will do the trick
- How about returning the result with optional looped path segments? (this might not come out that clean so I'll leave it up to you)